sdk/cosmos/azure_data_cosmos/src/query.rs

```diff
@@ -0,0 +1,157 @@
+/// #[derive(serde::Serialize)]
+/// struct CustomerInfo {
+///     id: u64,
+///     name: String
+/// }
+/// let query = Query::from("SELECT * FROM c WHERE ...")
+///     .with_parameter("@customer_info", CustomerInfo { id: 42, name: "Contoso".into() }).unwrap();
+/// ```
+#[derive(Debug, Serialize)]
+pub struct Query {
+    query: String,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")] // Don't serialize an empty array.
+    parameters: Vec<QueryParameter>,
+}
+
+impl Query {
+    /// Creates a new [`Query`] with the same text and the same parameters, with one additional parameter added.
+    ///
+    /// Returns an error if the value cannot be serialized to JSON.
+    pub fn with_parameter(
+        mut self,
+        name: impl Into<String>,
+        value: impl Serialize,
+    ) -> azure_core::Result<Self> {
+        let parameter = QueryParameter {
+            name: name.into(),
+            value: QueryParameterValue(serde_json::to_value(value)?),
+        };
+        self.parameters.push(parameter);
+
+        Ok(self)
+    }
+}
+
+impl<T: Into<String>> From<T> for Query {
+    fn from(value: T) -> Self {
+        let query = value.into();
+        Self {
+            query,
+            parameters: vec![],
+        }
+    }
+}
+
+/// Represents a single parameter in a Cosmos DB query.
+#[derive(Debug, Serialize)]
+struct QueryParameter {
+    name: String,
+    value: QueryParameterValue,
+}
+
+/// Represents a value that can be provided to a query parameter.
+#[derive(Debug, Serialize)]
+#[serde(transparent)]
+struct QueryParameterValue(serde_json::Value);
+
+#[cfg(test)]
+mod tests {
+    use serde::Serialize;
+
+    use crate::{NullValue, Query};
+
+    #[test]
+    pub fn serialize_query_without_parameters() {
+        let query: Query = "SELECT * FROM c".into();
+        let serialized = serde_json::to_string(&query).unwrap();
+        assert_eq!(serialized, r#"{"query":"SELECT * FROM c"}"#);
+    }
+
+    #[test]
+    pub fn serialize_query_with_string_parameters() {
+        let query = Query::from("SELECT * FROM c")
+            .with_parameter("name1", "value1")
+            .unwrap()
+            .with_parameter("name2", "value2")
+            .unwrap();
+        let serialized = serde_json::to_string(&query).unwrap();
+        assert_eq!(
+            serialized,
+            r#"{"query":"SELECT * FROM c","parameters":[{"name":"name1","value":"value1"},{"name":"name2","value":"value2"}]}"#
+        );
+    }
+
+    #[test]
+    pub fn serialize_query_with_various_parameter_types() {
+        #[derive(Serialize)]
+        struct ObjectParameter {
+            name: String,
+            value: String,
+        }
+        let obj_param = ObjectParameter {
+            name: "foo".into(),
+            value: "bar".into(),
+        };
+        let null_option: Option<&str> = None;
+
+        let query = Query::from("SELECT * FROM c")
+            .with_parameter("string_param", "value1")
+            .unwrap()
```

Comment:
Nit: would it be worth it just returning a result with a boxed error to avoid calling all these unwraps? This reads like a spoiled kid on Christmas.

Reply:
TIL that Rust tests can return `Result<T, E>` 😅 .